src/java/jni_helper_defines.h

```diff
@@ -163,31 +301,62 @@ static void ThrowNullPointerException(JNIEnv *jenv, const char *mesg) {
 }
 
 class ScopedMemory {
+#ifdef JAVA_DEBUG_TYPEMAPS
+    const static size_t beforeBytes = 32;
+    const static size_t afterBytes = 8192;
+#endif //JAVA_DEBUG_TYPEMAPS
+
 public:
-    ScopedMemory(size_t length) :
-            _memory(0), _length(length) {
-        _memory = calloc(length, 1);
+    ScopedMemory() {
+        set((size_t) 0);
     }
 
-    ScopedMemory(void *memory) :
-            _memory(memory), _length(0) {
+    ScopedMemory(size_t length) {
+        set(length);
     }
 
-    ScopedMemory(void *memory, size_t length) :
-            _memory(memory), _length(length) {
+    inline void set(size_t length, size_t size) {
+        set(length * size);
     }
 
-    ScopedMemory() :
-            _memory(0) {
-    }
+    inline void set(size_t length) {
+        _length = length;
+        if (0 == _length) {
+            _memory = 0;
+#ifdef JAVA_DEBUG_TYPEMAPS
+            _first = 0;
+            _last = 0;
+            _realPtr = _memory;
+#endif //JAVA_DEBUG_TYPEMAPS
+            return;
+        }
 
-    inline void set(void *memory) {
-        _memory = memory;
+#ifdef JAVA_DEBUG_TYPEMAPS
+        _realLen = beforeBytes+ length + afterBytes;
+        _realPtr = calloc(_realLen, 1);
+        _first = (void*) ((char*) _realPtr + beforeBytes);
+        _last = (void*) ((char*) _first + length);
+
+        memset(_realPtr, 0xDD, _realLen);
+        memset(_realPtr, 0xBB, beforeBytes);
+        memset(_first, 0, length);
+        memset((void*) ((char*) _last), 0xAA, beforeBytes);
```

Comment:
I think so.
This is actually copied from somewhere else, I'll have to go back and look.
It's copied back here to keep the 2 piles in sync in a very ugly way.